neqo-helper/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+    let socket = match UdpSocket::bind(local_addr) {
+        Err(e) => {
+            eprintln!("Unable to bind UDP socket: {}", e);
+            exit(1)
+        }
+        Ok(s) => {
+            // Don't let the host stack or network path fragment our IP packets
+            // (RFC9000, Section 14).
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpDontFrag, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6DontFrag, &true),
+            };
+            assert!(res.is_ok());
+            // Request IPv4 type-of-service (TOS) and IPv6 traffic class
+            // information for all incoming packets.
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpRecvTos, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6RecvTClass, &true),
+            };
+            assert!(res.is_ok());
+            // Request IPv4 time-to-live (TTL) and IPv6 hop count
+            // information for all incoming packets.
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpRecvTtl, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6RecvHopLimit, &true),
+            };
+            assert!(res.is_ok());
+            s
+        }
+    };
+    Ok(socket)
+}
+
+fn to_sockaddr(addr: SocketAddr) -> SockaddrStorage {
+    SockaddrStorage::from(addr)
+}
+
+pub fn emit_datagram(fd: i32, d: Datagram) -> io::Result<()> {
+    let iov = [IoSlice::new(&d[..])];
+    let tos = d.tos() as i32;
+    let ttl = d.ttl() as i32;
+    let cmsg_tos = match d.destination() {
+        SocketAddr::V4(..) => ControlMessage::IpTos(&tos),
+        SocketAddr::V6(..) => ControlMessage::Ipv6TClass(&tos),
+    };
+    let cmsg_ttl = match d.destination() {
+        SocketAddr::V4(..) => ControlMessage::IpTtl(&ttl),
+        SocketAddr::V6(..) => ControlMessage::Ipv6HopLimit(&ttl),
+    };
+    let sent = sendmsg(
+        fd,
+        &iov,
+        &[cmsg_tos, cmsg_ttl],
+        MsgFlags::empty(),
+        Some(&to_sockaddr(d.destination())),
+    )
+    .unwrap();
+    if sent != d.len() {
+        eprintln!("Unable to send all {} bytes of datagram", d.len());
+    }
+    Ok(())
+}
+
+fn to_socket_addr(addr: &SockaddrStorage) -> SocketAddr {
+    let fam = addr.family().unwrap();
+
+    let port = match fam {
+        AddressFamily::Inet => addr.as_sockaddr_in().unwrap().port(),
+        AddressFamily::Inet6 => addr.as_sockaddr_in6().unwrap().port(),
+        _ => todo!(),
+    };
+
+    match fam {
+        AddressFamily::Inet => SocketAddr::new(
+            std::net::IpAddr::V4(addr.as_sockaddr_in().unwrap().ip()),
+            port,
+        ),
+        AddressFamily::Inet6 => SocketAddr::new(
+            std::net::IpAddr::V6(addr.as_sockaddr_in6().unwrap().ip()),
+            port,
+        ),
+        _ => todo!(),
+    }
+}
+
+pub fn recv_datagram(
+    fd: i32,
```

Comment:
How often do we have a raw fd when calling this function?  Can this function call `to_raw_fd()` itself?

Reply:
I couldn't easily figure out how, because it's being called with a `mio::net::UdpSocket` from the server and a `std::net::UdpSocket` from the client. Will investigate.

Reply:
Oh, I missed that.  Unless the two implement a shared trait, you won't be able to.  But then it is possible to have that happen.  Whether it is worthwhile, I'll leave as an exercise :)

The basic idea here would be that you have two types that can feed in, so you would have a function signature that is generic:

```rust
pub fn recv_datagram<S: SocketTrait>(socket: S, ...) {
}
```

Then, to allow you to pass in either form, you would define that new trait with the functions you need as follows:

```rust
pub trait SocketTrait { // don't use my name
  fn raw_fd(&self) -> i32;
}
```

You can then implement that trait for the different types you want to support, like so:

```rust
impl SocketTrait for std::net::UdpSocket {
  fn raw_fd(&self) -> i32 { self.get_raw_fd() }
}


impl SocketTrait for mio::net::UdpSocket {
  fn raw_fd(&self) -> i32 { self.get_raw_fd() }
}
```

This means that your code can have multiple instantiations, one for each concrete type that implements this new trait.  In practice, you want one instantiation in the client and another in the server, so there is no code bloat.

This would make the interface more ergonomic, but it's not clear that we really need that capability in this instance.  I made the comment without realizing that there were different types involved.